impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -38,225 +82,312 @@ Join StrokePathGeometry::GetStrokeJoin() const {
-              forward + orientation * PathBuilder::kArcApproximationMagic,
-              forward);
-        }
-
-        vtx.position = position + orientation;
-        vtx_builder.AppendVertex(vtx);
-        vtx.position = position - orientation;
-        vtx_builder.AppendVertex(vtx);
-        std::vector<Point> arc_points;
-        arc.AppendPolylinePoints(scale, arc_points);
-        for (const auto& point : arc_points) {
-          vtx.position = position + point;
-          vtx_builder.AppendVertex(vtx);
-          vtx.position = position + (-point).Reflect(forward_normal);
-          vtx_builder.AppendVertex(vtx);
-        }
-      };
-      break;
+      return &StrokePathGeometry::CreateRoundCap;
     case Cap::kSquare:
-      cap_proc = [](VertexBufferBuilder<VS::PerVertexData>& vtx_builder,
-                    const Point& position, const Point& offset, Scalar scale,
-                    bool reverse) {
-        Point orientation = offset * (reverse ? -1 : 1);
-
-        VS::PerVertexData vtx;
-
-        Point forward(offset.y, -offset.x);
-
-        vtx.position = position + orientation;
-        vtx_builder.AppendVertex(vtx);
-        vtx.position = position - orientation;
-        vtx_builder.AppendVertex(vtx);
-        vtx.position = position + orientation + forward;
-        vtx_builder.AppendVertex(vtx);
-        vtx.position = position - orientation + forward;
-        vtx_builder.AppendVertex(vtx);
-      };
-      break;
+      return &StrokePathGeometry::CreateSquareCap;
   }
-  return cap_proc;
 }
 
-// static
-VertexBufferBuilder<SolidFillVertexShader::PerVertexData>
-StrokePathGeometry::CreateSolidStrokeVertices(
-    const Path& path,
-    Scalar stroke_width,
-    Scalar scaled_miter_limit,
-    const StrokePathGeometry::JoinProc& join_proc,
-    const StrokePathGeometry::CapProc& cap_proc,
-    Scalar scale) {
-  VertexBufferBuilder<VS::PerVertexData> vtx_builder;
-  auto point_buffer = std::make_unique<std::vector<Point>>();
-  // 512 is an arbitrary choice that should be big enough for most paths without
-  // needing to reallocate. If we have motivating benchmarks we should raise or
-  // lower this number, cause dnfield just made it up!
-  point_buffer->reserve(512);
-  auto polyline = path.CreatePolyline(scale, std::move(point_buffer));
+struct StrokeGenerator {
+  StrokeGenerator(const Path::Polyline& p_polyline,
+                  const Scalar p_stroke_width,
+                  const Scalar p_scaled_miter_limit,
+                  const StrokePathGeometry::JoinProc& p_join_proc,
+                  const StrokePathGeometry::CapProc& p_cap_proc,
+                  const Scalar p_scale,
+                  VertexWriter& p_vtx_builder)
+      : polyline(p_polyline),
+        stroke_width(p_stroke_width),
+        scaled_miter_limit(p_scaled_miter_limit),
+        join_proc(p_join_proc),
+        cap_proc(p_cap_proc),
+        scale(p_scale),
+        vtx_builder(p_vtx_builder) {}
```

Comment:
Here you are storing a raw pointer to vtx_builder, can you instead pass this in as an argument to `Generate`?

Reply:
Since this is an private class I think we can maintain this invaraint?

Reply:
It's kind of leaving a footgun in the code, seems like an easy change to remove it unless I'm missing something.

Reply:
How is it a footgun?

Reply:
Here's the official stance of the c++ style guide: https://google.github.io/styleguide/cppguide.html#Ownership_and_Smart_Pointers

This particular case is a grey area.  You have a non-owning raw pointer which is at risk of becoming a dangling pointer.

Here's one example where it can go wrong:
```c++
class Foo {
public:
  void good();
  Bar bad();
};

class Bar {
public:
  Bar(Foo& foo) : foo_(foo) {}
  Foo& foo_;
};

void Foo::good() {
  Bar bar(*this);
}

Bar Foo::bad() {
  return Bar(*this);
}

void bad2() {
  
}
```

You can imagine a few other ways this can go wrong.  We can try to avoid them, but that's why it's a foot gun and not a error.  There is no compile-time guarantee that you won't introduce a dangling pointer and that's of the intent of ownership section.

If you remove the field that is storing a raw pointer you remove the risk that it can become a dangling pointer.

At a bare minimum, any C++ class that has a raw pointer should define copy constructors which would remove many of the cases where you'd get dangling pointers.  I think just removing the field is easier option though.

Reply:
But that example isn't representative of what I've written. I've replaced closures closing over a reference with a class that contains that reference, which is more or less what the templated code could generate. The class is private and not long lived.

Reply:
> not long lived

Right, but there is no guarantee that it is not long lived, no guarantee that StrokeGenerator's lifetime is shorter than the VertexWriter and thus no guaranteed protection from a dangling pointer.

To make that guarantee you need:
1) remove the raw pointer field -- You can't have a dangling pointer if there is no pointer.
1) make that class not copiable or movable (less comprehensive) -- I think this gets rid of all the cases as well.

Reply:
I asked the google c++ community, here is the prevailing wisdom here:

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

```
Avoid defining functions that require a const reference parameter to outlive the call, because const reference parameters bind to temporaries. Instead, find a way to eliminate the lifetime requirement (for example, by copying the parameter), or pass it by const pointer and document the lifetime and non-null requirements.
```

In this case the function is the constructor of StrokeGenerator, so it is against the style recommendation.

go/totw/220 also has some expanded discussion on it.

Reply:
My read of both of those is to be careful with functions that return references, if the lifetime is potentially too short. but in this case I'm not returning a reference, I'm using the class as a storage for a reference. 

Happy to make it non-moveable and copyable though.

Reply:
> My read of both of those is to be careful with functions that return references,

The key text is "Avoid defining functions that require a const reference parameter to outlive the call".  The internal version is "Avoid defining functions that require a reference parameter to outlive the call."  That's what we have here with the constructor.  I'm happy to deviate from the style guide if we have a valid reason.  I think in this case you just prefer your style which I don't think is a valid justification, let me know if I'm misinterpreting =)

Reply:
The reference already outlives the call, and is used in a class as is. Its equivalent to closing over a non-const reference, which we do all the time. I'd argue that is actually less safe, since I could more easily call the closure in a scope where the reference had been destructed.

At any rate, it turns out its trivial to just pass the reference in so I've made this change.

Reply:
C++ style guide has guidance for lambda too: "Use default capture by reference ([&]) only when the lifetime of the lambda is obviously shorter than any potential captures."

I don't know if it is ever truly obvious in c++ haha.